crates/biome_js_analyze/src/lint/nursery/no_common_js.rs

```diff
@@ -0,0 +1,138 @@
+use biome_analyze::{context::RuleContext, declare_lint_rule, Rule, RuleDiagnostic, RuleSource};
+use biome_console::markup;
+use biome_js_syntax::{
+    global_identifier, JsCallExpression, JsReferenceIdentifier, JsStaticMemberAssignment,
+};
+use biome_rowan::{declare_node_union, AstNode};
+
+use crate::services::semantic::Semantic;
+
+declare_lint_rule! {
+    /// Disallow use of CommonJs module system in favor of ES6 style imports.
```

Comment:
We should change all the instances of "ES6 style imports" with "ESM", because that's the official name (right?). Ideally we should provide a couple of links for CommonJS and ESM for reference.

As for the rule itself, the rule should not be triggered if the manifest, the package.json, defines the project is a script (type field). The rule should not be triggered for .cjs files too.

Reply:
There are few cases:
File extension is .cjs,cts. Likely should allow the rule.
If package.json `type` field is empty or "commonjs".

But this might cause issues where this rule is most needed: projects using typescript/bundlers to bundle into "commonjs", but using mixed ESM/CommonJS syntax in source. If we disable rule on those 2 conditions automatically they will be left out. Should we add some sort of option, for those users?

I am almost leaning on ignoring package.json type field because of that, but not sure.

Reply:
We could start with your implementation, and then wait for possible feedback from the users.

Considering what you just explained to me, maybe we should expand the documentation and explain when users should or should not use the rule

Reply:
Disabled rule for cjs files and added bit of documentation.

Reply:
> Disabled rule for cjs files and added bit of documentation.

We should also disable the rule if the root is a `JsScript`. This allows supporting cases where `pckage.json` `type` field is explicitly set to `commomjs`. You can retrieve the root by using `ctx.root()` in `run`. You can take a look at the implementation of `noRedundantUseStrict`.

Reply:
That was what @ematipico suggested as well, but see my comment https://github.com/biomejs/biome/pull/3821#discussion_r1749251606 people who are migrating mixed codebases with bundlers/typescript will be not able to use this rule. If we disable this rule on commonjs packages by default, at very least we should provide an option to enable it, which I am not sure about.

Reply:
Oh sorry, I missed it. Looks good to me. Let's keep the rule as it is.